pallets/pool-registry/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+
+	#[pallet::storage]
+	#[pallet::getter(fn get_pool_metadata)]
+	pub(super) type PoolMetadata<T: Config> =
+		StorageMap<_, Blake2_256, T::PoolId, PoolMetadataOf<T>>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A pool was created.
+		Created {
+			pool_id: T::PoolId,
+			admin: T::AccountId,
+		},
+		/// A pool was updated.
+		Updated { pool_id: T::PoolId },
+		/// Pool metadata was set.
+		MetadataSet {
+			pool_id: T::PoolId,
+			metadata: BoundedVec<u8, T::MaxSizeMetadata>,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Invalid metadata passed
+		BadMetadata,
+		/// Pre-requirements for a TrancheUpdate are not met
+		/// for example: Tranche changed but not its metadata or vice versa
+		InvalidTrancheUpdate,
+		/// No metadata for the given currency found
+		MetadataForCurrencyNotFound,
+		/// No Metadata found for the given PoolId
+		NoSuchPoolMetadata,
+		/// The given tranche token name exceeds the length limit
+		TrancheTokenNameTooLong,
+		/// The given tranche symbol name exceeds the length limit
+		TrancheSymbolNameTooLong,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create a new pool
+		///
+		/// Initialise a new pool with the given ID and tranche
+		/// configuration. Tranche 0 is the equity tranche, and must
+		/// have zero interest and a zero risk buffer.
+		///
+		/// The minimum epoch length, and maximum NAV age will be
+		/// set to chain-wide defaults. They can be updated
+		/// with a call to `update`.
+		///
+		/// The caller will be given the `PoolAdmin` role for
+		/// the created pool. Additional administrators can be
+		/// added with the Permissions pallet.
+		///
+		/// Returns an error if the requested pool ID is already in
+		/// use, or if the tranche configuration cannot be used.
+		#[pallet::weight(T::WeightInfo::create(tranche_inputs.len().try_into().unwrap_or(u32::MAX)))]
+		#[transactional]
+		pub fn create(
+			origin: OriginFor<T>,
+			admin: T::AccountId,
+			pool_id: T::PoolId,
+			tranche_inputs: Vec<
+				TrancheInput<T::Rate, T::MaxTokenNameLength, T::MaxTokenSymbolLength>,
+			>,
+			currency: T::CurrencyId,
+			max_reserve: T::Balance,
+			metadata: Option<Vec<u8>>,
+		) -> DispatchResult {
+			T::PoolCreateOrigin::ensure_origin(origin.clone())?;
+
+			// First we take a deposit.
+			// If we are coming from a signed origin, we take
+			// the deposit from them
+			// If we are coming from some internal origin
+			// (Democracy, Council, etc.) we assume that the
+			// parameters are vetted somehow and rely on the
+			// admin as our depositor.
+			let depositor = ensure_signed(origin).unwrap_or(admin.clone());
+
+			match T::ModifyPool::create(
+				admin.clone(),
+				depositor,
+				pool_id,
+				tranche_inputs,
+				currency,
+				max_reserve,
+				metadata,
+			) {
+				Ok(_) => {
+					Self::deposit_event(Event::Created { pool_id, admin });
+					Ok(())
+				}
+				Err(e) => Err(e),
+			}
```

Comment:
`map(|_| Self::deposit_event(Event::Created{pool_id, admin}))` instead of match. Or was this not possible?

Reply:
Hm... doesn't this swallow the `Error`?

Reply:
No. See https://doc.rust-lang.org/src/core/result.rs.html#767

Reply:
Oh sweet!